src/libstore/filetransfer.cc

```diff
@@ -47,6 +49,56 @@ struct curlFileTransfer : public FileTransfer
     std::random_device rd;
     std::mt19937 mt19937;
 
+#if NIX_WITH_S3_SUPPORT
+public:
+    // Public method to clear credential provider cache (called from AWS CRT shutdown)
+    void clearCredentialCache();
+
+private:
+    struct CredentialProviderCache
+    {
+        // Key: profile name (empty string for default profile)
+        std::map<std::string, std::shared_ptr<AwsCredentialProvider>> providers;
+    };
+
+    Sync<CredentialProviderCache> credentialProviderCache;
+
+    /**
+     * Get or create a credential provider for the given profile.
+     * Thread-safe: holds lock for entire duration to prevent duplicate creation.
+     */
+    std::shared_ptr<AwsCredentialProvider> getOrCreateCredentialProvider(const std::string & profile)
+    {
+        auto cache(credentialProviderCache.lock());
+
+        // Check if provider exists
+        auto it = cache->providers.find(profile);
+        if (it != cache->providers.end()) {
+            return it->second;
+        }
+
+        debug(
+            "[pid=%d] creating new AWS credential provider for profile '%s'",
+            getpid(),
+            profile.empty() ? "(default)" : profile.c_str());
+        auto provider =
+            profile.empty() ? AwsCredentialProvider::createDefault() : AwsCredentialProvider::createProfile(profile);
+
+        auto sharedProvider = std::shared_ptr<AwsCredentialProvider>(std::move(provider));
+        cache->providers[profile] = sharedProvider;
+        return sharedProvider;
+    }
+
+    /**
+     * Invalidate a cached credential provider (e.g., on auth failure)
+     */
+    void invalidateCredentialProvider(const std::string & profile)
+    {
+        auto cache(credentialProviderCache.lock());
+        cache->providers.erase(profile);
+    }
+#endif
```

Comment:
All this stuff is effectively global; I would like it to move to a another file and not be part of `curlFileTranfer`. `curlFileTranfer` can just come it instead.